### PR TITLE
Fix NetworkWriter.AsyncFlushPages hanging FlushEvent waiters on send failure

### LINKED ISSUE #1929: Bug: Gossip timeout causes GarnetClient disposal cascade, ObjectDisposedException, and process unresponsiveness
### Describe the bug

In Garnet v1.1.10 cluster mode over a WAN link, a gossip round timeout triggers a cascade of failures that eventually makes the entire Garnet process unresponsive. The sequence is:

1. Gossip to the remote node times out (expected on WAN links)
2. The gossip `GarnetClient` is disposed
3. The underlying TCP socket is disposed, throwing `ObjectDisposedException` in `ClientTcpNetworkSender.SendResponse`
4. An internal throttle counter goes negative ("Throttle count is negative")
5. AUTH reconnect attempts throw `GarnetClientDisposedException` with 30+ nested reconnect calls in the stack trace
6. The Garnet main thread becomes blocked. Even `DBSIZE` (an O(1) counter read with no keyspace traversal) times out at 10 seconds over TCP.

This cascade starts from a single gossip timeout and takes approximately 1 hour to fully degrade the process to the point where client commands cannot be served. The process must be restarted to recover.

The root issue appears to be that the gossip `GarnetClient` is not resilient to transient network timeouts. Once disposed, the reconnect logic enters what looks like an infinite retry loop, consuming the main thread.

### Steps to reproduce the bug

1. Deploy Garnet v1.1.10 in cluster mode with 2 nodes (1 primary, 1 replica) across a WAN link with approximately 150ms RTT.

2. Configure with these cluster flags:

```ini
--cluster
--cluster-timeout 15
--gossip-delay 30
--cluster-announce-ip <public-ip>
```

3. Wait for a gossip round to time out. This occurs naturally on WAN links. The first sign in stdout is:

```log
warn: ClusterManager[0] GOSSIP round faulted
warn: ClusterManager[0] GOSSIP to remote node [<id> <ip>:6379] timeout!
```

4. Within the same gossip round, the cascade begins. The following sequence was captured from a production deployment:

```log
fail: ClusterManager[0] Exception calling networkSender.SendResponse in AsyncFlushPages
System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'System.Net.Sockets.Socket'.
   at System.Net.Sockets.Socket.SendAsync(SocketAsyncEventArgs e, CancellationToken cancellationToken)
   at Garnet.client.ClientTcpNetworkSender.SendResponse(Byte[] buffer, Int32 offset, Int32 count, Object context)
   at Garnet.client.NetworkWriter.AsyncFlushPages(Int64 fromAddress, Int64 untilAddress)

fail: ClusterManager[0] Exception calling networkSender.SendResponse in AsyncFlushPages
System.Exception: Throttle count is negative
   at Garnet.client.ClientTcpNetworkSender.SendResponse(Byte[] buffer, Int32 offset, Int32 count, Object context)
   at Garnet.client.NetworkWriter.AsyncFlushPages(Int64 fromAddress, Int64 untilAddress)
```

5. Client commands continue to be served for some time but the process degrades progressively. After approximately 1 hour, even trivial commands fail. Test by sending `DBSIZE` over a raw TCP connection:

```text
*2\r\n$4\r\nAUTH\r\n$<len>\r\n<password>\r\n
*2\r\n$6\r\nDBSIZE\r\n
```

This times out with no response from Garnet (TCP read timeout at 10 seconds). `DBSIZE` is an O(1) operation, so a timeout here indicates the main Garnet thread is blocked.

6. The only recovery is to restart the Garnet process. After restart, all commands respond normally again.

### Expected behavior

A transient gossip timeout should trigger a retry or backoff, not dispose the gossip client permanently. If the client must be disposed, the reconnect logic should not enter a nested loop of 30+ retry attempts. The main Garnet thread should remain responsive to client commands even when the gossip subsystem is degraded. A single gossip failure should not cascade into process-wide unresponsiveness requiring a restart.

### Screenshots

_No response_

### Release version

v1.1.10 (latest stable)

### IDE

_No response_

### OS version

Windows Server 2022 (both nodes)

### Additional context

**Cluster configuration used (identical on both nodes except role):**

```ini
--port 6379
--bind 0.0.0.0
--memory 8g
--aof
--aof-commit-freq 1000
--aof-size-limit 512m
--compaction-freq 3600
--pubsub-pagesize 1m
--cluster
--cluster-timeout 15
--gossip-delay 30
--cluster-replication-reestablishment-timeout 30
--cluster-announce-ip <public-ip>
```

**Environment details:**

- 2-node cluster: 1 primary, 1 replica
- Cross-region deployment, RTT approximately 150ms between nodes
- Garnet runs as a Windows service via NSSM
- StackExchange.Redis 2.12.14 client on .NET Framework 4.8
- Approximately 1 million keys at time of failure
- Neither node uses SSD storage tier (`--storage-tier` is not set)

**AUTH reconnect stack trace (30+ nested levels):**

```log
fail: ClusterManager[0] AUTH returned error
Garnet.client.GarnetClientDisposedException: Cannot access a disposed object.
Object name: 'GarnetClient'.
   at Garnet.client.GarnetClient.ThrowException(Exception e) in /_/libs/client/GarnetClient.cs:line 1128
   at Garnet.client.GarnetClient.InternalExecuteAsync(TcsWrapper tcs, Memory<byte> respOp, ...)
   at Garnet.client.GarnetClient.ExecuteForMemoryResultWithCancellationAsync(...)
   at Garnet.client.GarnetClient.ConnectAsync(CancellationToken token) in /_/libs/client/GarnetClient.cs:line 255
   at Garnet.client.GarnetClient.ReconnectAsync(CancellationToken token) in /_/libs/client/GarnetClient.cs:line 454
   at Garnet.client.GarnetClient.ExecuteForMemoryResultWithCancellationAsync(...)
   at Garnet.client.GarnetClient.ReconnectAsync(CancellationToken token)
   at Garnet.client.GarnetClient.ExecuteForMemoryResultWithCancellationAsync(...)
   at Garnet.client.GarnetClient.ReconnectAsync(CancellationToken token)
   ... (pattern repeats 30+ times: ReconnectAsync -> ExecuteForMemory -> ReconnectAsync)
```

The stack trace shows `ReconnectAsync` calling `ExecuteForMemoryResultWithCancellationAsync` which calls `ReconnectAsync` again in what appears to be an unbounded retry loop. The `CancellationToken` from the failed gossip operation may be contributing to this loop, similar in nature to the shared `CancellationTokenSource` issue described in [#1407](https://github.com/microsoft/garnet/issues/1407) but in the gossip subsystem rather than the replication subsystem.

**How we identified the main thread block:**

We sent raw RESP commands over TCP (no client library, just `TcpClient` in PowerShell) to eliminate the possibility of client-side issues:

```powershell
# PowerShell raw RESP test
$c = [System.Net.Sockets.TcpClient]::new('127.0.0.1', 6379)
$s = $c.GetStream(); $s.ReadTimeout = 10000
# AUTH
$auth = "*2`r`n`$4`r`nAUTH`r`n`$<len>`r`n<password>`r`n"
[Text.Encoding]::ASCII.GetBytes($auth).ForEach({ $s.Write($_, 0, $_.Length) })
# read AUTH response
$buf = [byte[]]::new(256); $s.Read($buf, 0, $buf.Length)
# DBSIZE
$cmd = "*2`r`n`$6`r`nDBSIZE`r`n"
[Text.Encoding]::ASCII.GetBytes($cmd).ForEach({ $s.Write($_, 0, $_.Length) })
# THIS TIMES OUT when Garnet is degraded -- exception: "Unable to read data from
# the transport connection: A connection attempt failed because the connected
# party did not properly respond..."
$buf2 = [byte[]]::new(8192); $s.Read($buf2, 0, $buf2.Length)
$c.Close()
```

`AUTH + PING` worked during early degradation (Garnet still processing commands). `AUTH + DBSIZE` timed out after the degradation progressed, meaning Garnet's main processing thread was blocked and could not drain the TCP receive buffer. `DBSIZE` is a simple counter read requiring no keyspace traversal.

This is the same class of problem as [#1407](https://github.com/microsoft/garnet/issues/1407) (shared state poisoned by a single failure) but in a different subsystem. In #1407 it is the replication `CancellationTokenSource`. In this case it appears to be the gossip `GarnetClient` and its reconnect loop.

### Fixes on main not backported to v1.1.x

There are at least 3 fixes merged to `main` that address production stability issues but were never backported to the v1.1.x stable line. v1.1.10 is the end of the v1.1.x series - no v1.1.11 exists as of July 2026. The comparison between v1.1.10 and main shows 194 commits across 1,192 changed files with no backport effort:

| PR | Fix | Backport Status |
| --- | --- | --- |
| [#1753](https://github.com/microsoft/garnet/pull/1753) | Transient X-lock leak after "Entry does not fit on page" (fixes [#1749](https://github.com/microsoft/garnet/issues/1749)). AOF entry exceeding AofPageSize throws TsavoriteException which unwinds before TransientXUnlock runs, leaving the hash bucket locked forever. Subsequent operations on the same bucket spin at 100% CPU. This is the same blocked-thread class of bug as the gossip cascade - a single exception poisons shared state. | Merged to `main`. **Not backported.** |
| [#1838](https://github.com/microsoft/garnet/pull/1838) | Fix AOF page/memory size defaults and validation (fixes [#1811](https://github.com/microsoft/garnet/issues/1811), [#1770](https://github.com/microsoft/garnet/issues/1770)). Old defaults had AofPageSize=4m but main-log PageSize=32m, so entries up to 32m could be committed to the main store but could not fit on an AOF page - triggering the X-lock leak from #1749. Bumps defaults and adds startup validation. | Merged to `main`. Backport PR [#1844](https://github.com/microsoft/garnet/pull/1844) targeting `release/v1` was **closed as draft, never merged.** |
| [#1305](https://github.com/microsoft/garnet/pull/1305) | Fix replica checkpointing divergence with CommitAOF calls. Also introduces the `ClusterReplicationReestablishmentTimeout` setting (which we use via `--cluster-replication-reestablishment-timeout 30`). | Merged to `main`. The checkpointing fix is main-only. The reestablishment flag was backported as part of the v1.1.x feature set. |

The Garnet team appears to have shifted all development to the v2.0.x beta series (15 beta releases since December 2025). No patch releases are planned for v1.1.x. This leaves production users on the stable line running with known bugs that are already fixed on main. If the gossip cascade fix is straightforward (adding null/connectivity checks and a retry with backoff in the gossip path), a v1.1.11 patch would allow production users to stay on the stable line until v2.0.x is production-ready.

### PR DESCRIPTION
**Pull request by hexonal**
## Context

This is a follow-up on the "main thread blocking / hang" symptom described in #1929. #1929 traces back to a `NullReferenceException` in `GarnetServerNode.TryClusterPublish` after a gossip timeout, which is fixed by the (currently unmerged) maintainer PR #1931 via an `!IsConnected` guard.

**This PR does not touch that code path and does not claim to resolve #1929 on its own.** It fixes a separate bug that is reachable from the same general area (a broken/disposed cluster connection during a flush) and produces the same class of symptom — a caller thread hanging indefinitely instead of failing fast. Whether or not it is *the* root cause of every hang reported under #1929 I can't say with confidence; what I can say precisely is described below.

## Root cause

`NetworkWriter.AsyncFlushPages` sends each buffered page via `networkSender.SendResponse(...)`, passing a `PageAsyncFlushResult` as the completion context. Normally, `SendResponse` (synchronously or via a completion callback) eventually invokes `AsyncFlushPageCallback(context)`, which decrements `PageAsyncFlushResult.count.count` and, once it reaches zero, calls `FlushEvent.Set()`.

If `SendResponse` throws (e.g. because the underlying connection was concurrently disposed/broken — which is exactly what happens when a gossip timeout tears down a cluster connection), the existing code:

```csharp
catch (Exception ex)
{
    logger?.LogError(ex, "Exception calling networkSender.SendResponse in AsyncFlushPages");
    networkHandler.Dispose();
}
```

logs the error and disposes the handler, but never invokes `AsyncFlushPageCallback` for that page. `FlushEvent` is therefore never signaled for this flush. Any thread blocked on it — most notably `GarnetClient.InternalExecuteNoResponse`, which `flushEvent.Wait()`s synchronously and is used by cluster PUBLISH forwarding — hangs indefinitely (or until an external timeout/cancellation, if the caller supplied one).

## Fix

Explicitly invoke `AsyncFlushPageCallback(asyncResult)` in the catch block so the completion counter is still decremented and `FlushEvent` still gets signaled when a send fails, matching what would have happened on a successful send. 3-line change in `libs/client/NetworkWriter.cs`.

## Known related limitation (not fixed here)

`ClientTcpNetworkSender.SendResponse` can complete synchronously and inline (when the underlying `socket.SendAsync` returns `false`), which means `AsyncFlushPageCallback` can already be running inline underneath `SendResponse` when it throws. In the narrow case where `AsyncFlushPageCallback` itself throws (e.g. from `epoch.BumpCurrentEpoch`, reachable from inside it), this fix's catch block would invoke `AsyncFlushPageCallback` a second time for a result whose completion counter was already decremented once, which could corrupt `count.count` for that flush batch. I have not observed this in practice and did not reproduce it — flagging it as a known, narrow, pre-existing-adjacent edge case rather than silently leaving it unaddressed. Happy to follow up separately if it's judged worth guarding against explicitly.

## Testing

- Added `AsyncFlushPagesSignalsFlushEventOnSendFailure` in `test/standalone/Garnet.test/GarnetClientTests.cs`: parks a thread on `flushEvent.Wait()` (mirroring a caller already blocked when the connection dies), swaps in a network sender stub whose `SendResponse` always throws, then triggers `AsyncFlushPages` and asserts the waiter is released rather than timing out.
  - Verified the test reproduces the bug: with the 3-line fix reverted, the test fails with `OperationCanceledException` after a 3s cancellation timeout (a real hang). With the fix restored, it passes in under a second.
  - The test needs `InternalsVisibleTo` on `Garnet.client` for the test assembly (`NetworkWriter` is `internal sealed`); added to `libs/client/Garnet.client.csproj` using the repo's existing signing key (same key already used by `libs/host/Garnet.host.csproj` and others).
- Isolated test run: 15/15 passes.
- Full `GarnetClientTests` class (20 tests): 8/8 full-class runs green, no flakes observed.
- `dotnet build` succeeds for `Garnet.client.csproj`, `Garnet.cluster.csproj`, and `Garnet.test.csproj`.
- `dotnet format --verify-no-changes` is clean on the changed files.
- I did not reproduce the original cluster/gossip-timeout scenario from #1929/#1931 end-to-end; testing here is at the `NetworkWriter` unit level described above, not a full cluster integration repro.

### CODE REVIEW

**Comment by hexonal** [comment]
One red job on the last run: `Garnet Standalone (ubuntu-latest, net10.0, Debug, Garnet.test)` — not a test failure, the whole `Run tests Garnet.test` step timed out after 45 minutes. Log shows this PR's own new test (`AsyncFlushPagesSignalsFlushEventOnSendFailure`) passed fine early on (16ms); the actual hang happened later, after `RespAdminCommandsTests.SeSaveRecoverMultipleKeysTest("16k","16k")` completed, with no further progress logged for 44 minutes before the timeout. Looks like the same class of intermittent persistence-test hang addressed before in #1932, unrelated to this PR's diff.

**Comment by hexonal** [comment]
Update: filed #1950 — turns out this test has a real, reliably-reproducible bug (not just flakiness), likely related to the recovery/eviction interaction #1932 touched. See there for repro steps and analysis.

File: libs/client/NetworkWriter.cs [hunk 1]
```diff
@@ -333,6 +333,9 @@ public void AsyncFlushPages(long fromAddress, long untilAddress)
                     {
                         logger?.LogError(ex, "Exception calling networkSender.SendResponse in AsyncFlushPages");
                         networkHandler.Dispose();
+                        // Still signal completion for this page so FlushEvent waiters (e.g. GarnetClient
+                        // callers blocked in InternalExecuteNoResponse) are not left hanging indefinitely.
+                        AsyncFlushPageCallback(asyncResult);
```
[the remark this answers is not in the document]
Comment (vazois) [reply libs/client/NetworkWriter.cs:338]: @hexonal, this seems serious. Can we guarantee that AsyncFlushPages does not decrement the value twice?
In the worst case, it might put the count into a state that creates the same issue that you were trying to resolve.

[the remark this answers is not in the document]
Comment (hexonal) [reply libs/client/NetworkWriter.cs:338]: I went through this carefully because it's a fair concern. `AsyncFlushPageCallback` does `Interlocked.Decrement(ref result.count.count)`, and only signals `FlushEvent` once that hits zero.

Looking at `ClientTcpNetworkSender.SendResponse`, every exception it can throw falls into one of two buckets: either it happens before the `SocketAsyncEventArgs` is even dispatched (throttle wait, checkout, buffer setup — no callback has run yet), or `socket.SendAsync()` itself throws synchronously, which per the SAEA contract means the operation was never actually queued, so `Completed` can never fire for it afterward. If `SendAsync` succeeds instead, `SendResponse` just returns normally and the catch block never runs for that page at all.

So there's no path where a page gets counted twice — the callback fires from exactly one place per page, either the normal completion or (only when the send never got queued) the new catch-block call, never both. I didn't just reason through this abstractly, I diffed the actual send path line by line against `main` to make sure I wasn't missing a partial-completion case.

File: test/standalone/Garnet.test/GarnetClientTests.cs [hunk 2]
```diff
@@ -568,5 +571,106 @@ public async Task MultipleSocketPing([Values] bool useTls)
...
+                // main thread has a deterministic (bounded-spin, not fixed-sleep) signal that
+                // the waiter is about to enter its wait rather than merely having started.
+                var aboutToWait = false;
+                Exception waiterException = null;
+                var waiterThread = new Thread(() =>
+                {
+                    Volatile.Write(ref aboutToWait, true);
+                    try { flushEvent.Wait(cts.Token); }
+                    catch (Exception ex) { waiterException = ex; }
+                })
```
[the remark this answers is not in the document]
Comment (hexonal) [reply test/standalone/Garnet.test/GarnetClientTests.cs]: Good catch, fixed in c4e561df — you're right that flipping `aboutToWait` before actually calling `Wait()` still left a real window.

Went one step further than the suggested fix, though: instead of registering the wait via `WaitAsync()` on a background thread and blocking on the returned task there, I realized there's no need for a separate thread at all. `WaitAsync()` registers with the underlying `SemaphoreSlim` synchronously, before it returns — so the test can just call `flushEvent.WaitAsync(cts.Token)` on the main thread, trigger the failing flush right after (no rendezvous needed, since registration already happened), and then block on the returned `Task` with a timeout. That removes the `aboutToWait` flag, the `SpinWait`, and the background thread entirely, not just the race between them.

Ran it 5x in a row locally plus the full `GarnetClientTests` suite (20/20) to sanity-check.

**Comment by vazois** [comment]
@hexonal, when this gets merged are you going to back-port it?

**Comment by hexonal** [comment]
Happy to -- checked `release/v1` directly and it has the exact same gap: the `catch` block in `AsyncFlushPages` logs and disposes the network handler but never calls `AsyncFlushPageCallback`, so it'd hang the same way on that branch.

I'll open the backport PR once this one merges into `main` (so it can reference a landed commit rather than a moving target). Let me know if you'd rather I do it in parallel instead.

**Review by vazois — APPROVED** [review]

**Comment by vazois** [comment]
@hexonal, are you planning to backport this?

**Comment by hexonal** [comment]
Yes -- sorry for the delay, this one landed while I was away from it. Opened as #1999.

Confirmed the gap is really still there on `release/v1` rather than assuming it: that branch's `AsyncFlushPages` catch block is character-for-character what `main`'s was before this PR.

It's a hand-applied backport rather than a cherry-pick, since the test file is at `test/Garnet.test/` on v1 (vs `test/standalone/Garnet.test/` on main) and v1's fixture derives from `AllureTestBase`. The production diff is byte-identical to what landed here; the only difference in the test is two doc-comment sentences reworded to drop stale references to an earlier iteration of it from this PR's review -- test logic is unchanged. I diffed `INetworkSender`, `CompletionEvent`, and both ctors across the 245 commits between the branches first to be sure the test would port cleanly.

Verified on v1 that the test actually catches the bug there: reverted just the `NetworkWriter.cs` hunk and it fails (3s timeout, waiter never released), restored it and it passes. Details and the local-verification limits are in #1999's description.
